New diagnostic capabilities in phoebuspy.py and Phoebus

curiousmiah opened the pull request:
## PR Summary

This PR provides new diagnostic tools in phoebuspy.py.  In particular it includes capabilities to plot and make movies of slices of 3D simulations.  It includes 2D slices and 1D slices.  In addition, there was a bug in Phoebus when including flux_divergence and src_terms in the output.  This has been fixed.

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [ ] Format your changes by calling `scripts/bash/format.sh`.
- [ ] Explain what you did.
- [ ] Make any necessary changes to the documentation.

Commit:
adding Dump3D class and starting capability to PlotScatter3DvsRadius

Commit:
Added CalcScattervsRad to phoebuspy.py

Commit:
making ScatterVsRad 20 times faster

Commit:
Added ability to save OneDProfile data.

Commit:
Starting to add Make2DSlice capability

Commit:
First working draft of Make2DSlices

Commit:
Added first draft of Movie2DSlices

Commit:
Added Parameter file to phoebuspy.py

Commit:
fixing some bugs in Movie2DSlices

Commit:
Added MakeSlicesVsTime and MakeSlicesVsCoord phoebuspy.py

Commit:
fixing indexing of src within the set_flux_src_diag code.

Commit:
finalizing 1D slice movie capabilities

Commit:
fixing formatting in fluid.cpp

Yurlungur commented on src/fluid/fluid.cpp:
```diff
@@ -645,11 +645,11 @@ TaskStatus CalculateFluidSourceTerms(MeshData<Real> *rc, MeshData<Real> *rc_src)
-        src(diag::src_terms(2), b, k, j, i) = src(c::momentum(1), k, j, i);
-        src(diag::src_terms(3), b, k, j, i) = src(c::momentum(2), k, j, i);
-        src(diag::src_terms(4), b, k, j, i) = src(c::energy(), k, j, i);
+        src(b, diag::src_terms(0), k, j, i) = 0.0;
+        src(b, diag::src_terms(1), k, j, i) = src(b, c::momentum(0), k, j, i);
+        src(b, diag::src_terms(2), k, j, i) = src(b, c::momentum(1), k, j, i);
+        src(b, diag::src_terms(3), k, j, i) = src(b, c::momentum(2), k, j, i);
+        src(b, diag::src_terms(4), k, j, i) = src(b, c::energy(), k, j, i);
```
good catch!

curiousmiah replied:
Thanks ;)

Yurlungur commented on scripts/python/phoebuspy.py:
```diff
@@ -59,6 +115,82 @@ def __init__(self, filename):
                 pass
         return
 
+def ReadParameterFile(FileName='Params.dat'):
```
As I understand it, this is a parser for your script, right? It might be easier/more maintainable to use `pyyaml` and yaml?

curiousmiah replied:
Good point!  I'm clearly "old school".

curiousmiah replied:
I just changed the ReadParameterFile file to use yaml.

Yurlungur commented on scripts/python/phoebuspy.py:
```diff
@@ -131,6 +263,525 @@ def MovieGR(data,varname='monopole_gr/rhoadm',anax=None,anay=None,ylim=None,xlim
+    ## Stack for plotting
+    #varpercentiles = np.vstack([p20, p50, p80])
+    
+    varpercentiles = np.zeros((3,nradbins))
+    varhopper = [[] for _ in range(nradbins)]
+    for im in range(data.NumMB):
```
is this really a scatter plot or is it a historgram? If it's the latter, we might want to rename it. Also, I imagine there's some way to get matplotlib to do this automatically

curiousmiah replied:
This plots an "average" radial profile for 3D sims.  For example, it plots something like <rho> vs. radius.  It's not really a histogram.  To create this "average" radial profile, I need to calculate some sort of statistic as a function of radius, hence the binning.  Instead of an average, I calculate the median.  But I also plot the 20 and 80th percntiles vs. radius as well.  This will give a sense of median and variation as a function of radius.  While I'm using statistics, I wouldn't call these histograms.  Maybe I could call it CalcStatistics3DvsRadius?

Yurlungur replied:
sure that works

Yurlungur commented on scripts/python/phoebuspy.py:
```diff
@@ -131,6 +263,525 @@ def MovieGR(data,varname='monopole_gr/rhoadm',anax=None,anay=None,ylim=None,xlim
+        # Find zones where the slice falls between w_local edges
+        mask = (w_local[:-1] <= slice) & (slice < w_local[1:])
+        
+        if (np.any(mask)):
+            assert np.sum(mask) == 1, f"Multiple indices match in meshblock {im}"
```
I'm trying to understand the implication of the slicing machinery here. The cell that overlaps with the axis is chosen, is that right?

curiousmiah replied:
Yes.  w_local is an array of the edges for the axis under consideration.  In the case that two overlap the axis -- or more specifically the slice -- then it picks the one where the lower edge is on the axis (or slice).

Yurlungur replied:
gotcha. Yeah that seems reasonable

Yurlungur commented on scripts/python/phoebuspy.py:
```diff
@@ -131,6 +263,525 @@ def MovieGR(data,varname='monopole_gr/rhoadm',anax=None,anay=None,ylim=None,xlim
+
+    slice_data = []
+    
+    for im in range(data.NumMB):
```
Can this loop be "numpy-ized" so that there doesn't need to be an explicit for loop?

curiousmiah replied:
Maybe.  I was trying, but the conditionals, the construction of the dicts for each block, and the variable sizes of the vars makes it difficult to numpy-ize.  ChatGPT wasn't much help in this regard either.

Yurlungur replied:
ok its fine for now

Yurlungur commented on scripts/python/phoebuspy.py:
```diff
@@ -131,6 +263,525 @@ def MovieGR(data,varname='monopole_gr/rhoadm',anax=None,anay=None,ylim=None,xlim
+
+    return slice_data
+
+#def Make1DSlice(data,keepaxis=1,slice=[0.,0.],extractvars=['p.density']):
```
There is quite a bit of commented out code throughout the file. Do you need/want to keep it all? Looks like quite a bit of it is previous drafts or unused debugging statements, etc.

curiousmiah replied:
Good point.  I'll clean it up!

curiousmiah replied:
Done.

Yurlungur reviewed (approved):
This is a welcome addition! A few nitpicks and suggestions below. But it's all optional, we can merge when you're happy with it.

curiousmiah committed:
Changing name of CalcScatter3DvsRadius to CalcStatistics3DvsRadius

curiousmiah committed:
removing unneeded comments

Commit:
cleaned up commented code.

Commit:
using yaml syntax for the new runtime parameter file Params.yml

curiousmiah commented:
@Yurlungur , I've addressed all of the suggestions. I'm fairly certain that I have adopted all suggestions.  Please check to make sure that you like the direction of the changes.  If so, I'm ready for you to approve merge if all looks good to you as well.

Yurlungur commented on scripts/python/phoebuspy.py:
```diff
@@ -1,10 +1,23 @@
 import re
 from glob import glob
 
+#On Polaris, you need to create a conda environment via...
+#If phoebuspy conda environment is already active
+#conda deactivate
+#If you want to delete the old environment and start from scratch.
+#module unload xalt
+#conda env remove -n phoebuspy
+#If you want to create the phosbuspy conda environment
+#conda create -n phoebuspy numpy numexpr matplotlib h5py scipy pyyaml -y
+#The above needs to be once to define the environment...
+#Do this every time you need to use python on Polaris
+#conda activate phoebuspy
```
good to document! :+1:

Yurlungur reviewed (approved):
Looks good! Thanks @curiousmiah !
